specs/interop.md

```diff
@@ -0,0 +1,540 @@
+}
+```
+
+### CrossL2Inbox
+
+Address: `0x4200000000000000000000000000000000000022`
+
+The `CrossL2Inbox` is responsible for executing a cross chain message on the destination chain.
+It is permissionless to execute a cross chain message on behalf of any user. Certain protocol
+enforced invariants must be preserved to ensure safety of the protocol.
+
+#### Invariants
+
+- The timestamp of executing message MUST be less than or equal to the timestamp of the initiating message
+- The chain id of the initiating message MUST be in the dependency set
+- The executing message MUST be initiated by an externally owned account such that the top level EVM call frame enters the `CrossL2Inbox`
+- The chain id of the initiating message MUST not be the chain id of the executing message
+
+##### Only EOA Invariant
+
+The `onlyEOA` invariant on executing a cross chain message enables static analysis on executing messages.
+This allows for the derivation pipeline and block builders to reject executing messages that do not
+have a corresponding initiating message.
+
+It may be possible to relax this invariant in the future if the block building process is efficient
+enough to do full simulations to gain the information required to verify the existence of the
+initiating transaction. This would be done by adding an additional function to the `CrossL2Inbox`
+in the future. One possible way to handle explicit denial of service attacks is to utilize identity
+in iterated games such that the block builder can ban identities that submit malicious transactions.
+
+#### Executing Messages
+
+All of the information required to satisfy the invariants MUST be included to or committed to in the calldata
+of the function that is used to execute messages. Both the block builder and the smart contract use
+this information to ensure that all system invariants are held.
+
+```solidity
+function relayMessage(bytes32 _hash, CrossChainMessage memory _msg) public {
+    require(msg.sender == tx.origin);
+    require(_hash == hashCrossChainMessage(_msg));
+    require(_msg.timestamp <= block.timestamp);
+    require(_msg.chainid != block.chainid);
+    require(L1Block.isInInteropSet(_msg.chainid));
+
+    bool success = SafeCall.call({
+      _target: _msg.target,
+      _calldata: _msg.data
+    });
+}
+```
+
+By including the hash in the calldata, it makes static analysis much easier for block builders.
+Without the check onchain, the block builder would need to hash the `CrossChainMessage` to build the
+`eth_getLogs` query. With the hash onchain, the block builder can directly trust the hash.
+
+### L2ToL2CrossDomainMessenger
+
+Address: `0x4200000000000000000000000000000000000024`
+
+The `L2ToL2CrossDomainMessenger` is a higher level abstraction on top of the `CrossL2Outbox` that
+provides features necessary for secure transfers of `ether` and ERC20 tokens between L2 chains.
```

Comment:
As of now, the `L2ToL2CrossDomainMessenger` handles ETH natively, but ERC20s will be built on top of another layer of abstraction above that (like the StandardBridge)